build-system/babel-plugins/babel-plugin-transform-function-declarations/index.js

```diff
@@ -0,0 +1,154 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const BAIL_OUT_CONDITIONS = {
+  // If this isn't a FunctionDeclaration, bail out on modification.
+  isNotFunction: (t, path) => !t.isFunctionDeclaration(path.node),
+
+  // If this FunctionDeclaration is a Generator, bail out on modification.
+  isGenerator: (t, path) => path.node.generator,
+
+  // If this FunctionDeclaration doesn't have a single return statement, bail out on modification
+  isNotSingleReturnStatment: (t, path) => {
+    if (!path.get('body').isBlockStatement()) {
+      return false;
+    }
+    const body = path.get('body.body.0');
+    return !body || !body.isReturnStatement();
+  },
+
+  // Since we don't know if exported members are 'new'd, bail out on modification.
+  isExported: (t, path) =>
+    path.findParent(
+      path =>
+        path.isExportNamedDeclaration() || path.isExportDefaultDeclaration()
+    ),
+
+  // If the function contains usage of `arguments`, bail out on modification.
+  containsArgumentsUsage: (t, path) => {
+    let containsArgumentsUsage = false;
+    path.traverse({
+      Identifier(path) {
+        if (t.isIdentifier(path.node, {name: 'arguments'})) {
+          containsArgumentsUsage = true;
+          path.stop();
+        }
+      },
+    });
+    return containsArgumentsUsage;
+  },
+
+  // If the function contains a ThisExpression, converting from a FunctionDeclaration to a
+  // VariableDeclaration => VariableDeclarator => ArrowFunctionExpression isn't necessarily valid.
+  // Bail out on modification.
+  containsThisExpression: (t, path) => {
+    let containsThis = false;
+    path.traverse({
+      ThisExpression() {
+        containsThis = true;
+        path.stop();
+      },
+    });
+    return containsThis;
+  },
+
+  // If the FunctionDeclaration identifier is newed in the scope of this program, bail out on modification.
+  isNewedInProgramScope: (t, path) => {
+    const {name} = path.get('id').node;
+    let isNewed = false;
+    path
+      .findParent(path => path.isProgram())
+      .traverse({
+        NewExpression(path) {
+          if (t.isIdentifier(path.node.callee, {name})) {
+            isNewed = true;
+            path.stop();
+          }
+        },
+      });
+    return isNewed;
+  },
+};
+
+function createVariableDeclaration(t, path) {
+  const {params, body, async, id} = t.cloneNode(path.node);
+  const arrowFunction = t.arrowFunctionExpression(
+    params,
+    body.body[0].argument,
+    async
+  );
+  const declarations = t.variableDeclarator(
+    t.identifier(id.name),
+    arrowFunction
+  );
+
+  // Making an existing FunctionDeclaration a const VariableDeclaration is safe.
+  // The name was already reserved for the FunctionDeclaration.
+  return t.variableDeclaration('const', [declarations]);
```

Comment:
Do you still want to do this given the Chrome thing we were discussing the other day w/r/t let/const?

Reply:
Good question. For reference, we discussed the performance impact of let/const with TDZ on JavaScript Core.

Our current state is JavaScript Core devices are not anywhere near CPU bound for the work in this library. We should do some additional investigation though.